File: problems/0004_median-of-two-sorted-arrays/binary-search/solution.go
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
    A, m, B, n := nums1, len(nums1), nums2, len(nums2)
    if m > n {
        A, m, B, n = B, n, A, m
    }
    total := m + n
    h, isOdd := total / 2, total % 2 == 1
    iBegin, iEnd := 0, m
    for iBegin <= iEnd {
        i := (iBegin + iEnd) / 2
        j := h - i
        if i < iEnd && 0 < j && A[i] < B[j - 1] {
            iBegin = i + 1
        } else if iBegin < i && j < n && B[j] < A[i - 1] {
            iEnd = i - 1
        } else {
            minGreater := 0
            if i == m {
                minGreater = B[j]
            } else if j == n {
                minGreater = A[i]
            } else {
                if A[i] < B[j] {
                    minGreater = A[i]
                } else {
                    minGreater = B[j]
                }
            }
            if isOdd {
                return float64(minGreater)
            }
            maxLesser := 0
            if i == 0 {
                maxLesser = B[j - 1]
            } else if j == 0 {
                maxLesser = A[i - 1]
            } else {
                if A[i - 1] > B[j - 1] {
                    maxLesser = A[i - 1]
                } else {
                    maxLesser = B[j - 1]
                }
            }
            return float64(maxLesser + minGreater) / 2.0
        }
    }
    panic("Should NOT be HERE!")
}

File: problems/0004_median-of-two-sorted-arrays/binary-search/Solution.rs
use std::cmp;

impl Solution {
    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        let (A, B) = if nums1.len() <= nums2.len() {
                         (nums1, nums2)
                     } else {
                         (nums2, nums1)
                     };
        let (m, n) = (A.len(), B.len());
        let total = m + n;
        let (h, is_odd) = (total / 2, total % 2 == 1);
        let (mut iBegin, mut iEnd) = (0, m);
        while iBegin <= iEnd {
            let i = (iBegin + iEnd) / 2;
            let j = h - i;
            if i < iEnd && 0 < j && A[i] < B[j - 1] {
                iBegin = i + 1;
            } else if iBegin < i && j < n && B[j] < A[i - 1] {
                iEnd = i - 1;
            } else {
                let minGreater = if i == m {
                                     B[j]
                                 } else if j == n {
                                     A[i]
                                 } else {
                                     cmp::min(A[i], B[j])
                                 };
                if is_odd {
                    return minGreater as f64;
                }
                let maxLesser = if i == 0 {
                                    B[j - 1]
                                } else if j == 0 {
                                    A[i - 1]
                                } else {
                                    cmp::max(A[i - 1], B[j - 1])
                                };
                return (maxLesser + minGreater) as f64 / 2.0;
            }
        }
        panic!("Should NOT be HERE!");
    }
}
